### PR TITLE
ROX-26605: return error on matching fail for delegated ad-hoc scans

### PR DESCRIPTION
**Pull request by dcaravel**
### Description

Fixes an issue with roxctl (or API) scans that are delegated to a secured cluster.  When the metadata pull + indexing succeeds but matching fails the output appears to be a 'successful' scan - however `.scan` will be missing from the scan result (will include metadata only).

Now an error will be returned.

Also adds logging the delegated scanning request ID to the error so when examining administrative events or central logs we can differentiate between ad-hoc scans and non.

### User-facing documentation

- [x] CHANGELOG is updated **OR** update is not needed
- [x] [documentation PR](https://spaces.redhat.com/display/StackRox/Submitting+a+User+Documentation+Pull+Request) is created and is linked above **OR** is not needed

### Testing and quality

- [x] the change is production ready: the change is GA or otherwise the functionality is gated by a feature flag
- [x] CI results are inspected

#### Automated testing

No additional tests added.

#### How I validated my change

Deployed ACS 4.5.3 and while Scanner V4 was still loading vulns delegated a scan.  

This scan 'SHOULD' return an error instead of a 'partial' response:
```
$ rctl image scan -f --image=nginx --cluster=remote | jq ".name,.metadata.dataSource,.notes,.scan"
{
  "registry": "docker.io",
  "remote": "library/nginx",
  "tag": "latest",
  "fullName": "docker.io/library/nginx:latest"
}
{
  "id": "10d3b4dc-8295-41bc-bb50-6da5484cdb1a",
  "name": "Public DockerHub"
}
[
  "MISSING_SIGNATURE"
]
null
```
After the change, notice the scans will now return an error
```
$ oc set image deploy/central *=quay.io/rhacs-eng/main:4.6.x-742-gdf55853f2a

$ rctl image scan -f --image=nginx --cluster=remote | jq ".name,.metadata.dataSource,.notes,.scan"
ERROR:	Scanning image failed: could not scan image: "nginx": rpc error: code = FailedPrecondition desc = error delegating scan to cluster "b1a9eb99-4359-45d2-991f-6f05ebfac2b2" for image "docker.io/library/nginx:latest": retrieving image vulnerabilities from Scanner V4 [scannerv4]: get vulnerability report (reference: "index.docker.io/library/nginx@sha256:367678a80c0be120f67f3adfccc2f408bd2c1319ed98c1975ac88e750d0efe26"): get vulns: rpc error: code = FailedPrecondition desc = the matcher is not initialized: initial load for the vulnerability store is in progress. Retrying after 3 seconds...

```

### CODE REVIEW

File: central/image/service/service_impl.go [hunk 1]
```diff
@@ -568,11 +568,12 @@ func (s *serviceImpl) EnrichLocalImageInternal(ctx context.Context, request *v1.
 					// to help troubleshoot potential image name or caching issues log the request's image
 					// name as well.
 					logging.String("request_image", request.GetImageName().GetFullName()),
+					logging.String("request_id", request.GetRequestId()),
 				)
 
-				if imgExists {
-					// In case we hit an error during enriching, and the image previously existed, we will _not_ upsert it in
-					// central, since it could lead to us overriding an enriched image with a non-enriched image.
+				if imgExists || request.GetRequestId() != "" {
```
Comment (jvdm) [inline central/image/service/service_impl.go:574]: Help me understand this: It's possible that the metadata was retrieved successfully, but since this is "ad-hoc" (ie. API call driven scan), we are OK?

Also, wondering if we need to check for API-call drive scans for signatures as well (line 587).

Comment (dcaravel) [reply central/image/service/service_impl.go:574]: > Help me understand this: It's possible that the metadata was retrieved successfully, but since this is "ad-hoc" (ie. API call driven scan), we are OK?

Right, for example: user executes `roxctl image scan ...`, the request is delegated to the secured cluster, metadata retrieved successfully, image indexed successfully, matching request sent to Central but fails due to vuln data not being loaded yet. Current state a 'successful' response is returned to the API/roxctl, but when the JSON is inspected the scan is empty. This is wrong and an error should be returned - otherwise CI or other systems will believe there are no vulnerabilities detected and the image is clean.

> Also, wondering if we need to check for API-call drive scans for signatures as well (line 587).

Good callout - my first reaction is to leave it be until there is a reason to change it + given how close to release this is and being unclear how to test the failure scenario.

Currently in Sensor no [error is returned when sig pulling fails](https://github.com/stackrox/stackrox/blob/4575c4a00c790f13c41226c952968ab2357d245a/sensor/common/scan/scan.go#L378-L401), if 'continuing' that behavior we shouldn't error out in Central either when there are sig validation errors. That being said the chance of an error occurring for sig validation is much lower since there are no RPC calls being made to validate signatures.  On the flip side Central enrichments will fail if there is a sig validation error - so the two are not in sync currently.  We did have an effort starting to revisit the enrichers because there are many nuances where the behaviors are not the same.

I believe the way delegated scanning is currently working makes 'more' sense because a `roxctl image scan` probably shouldn't error out when sig validation fails (is no reason to), however a `roxctl image check` should fail  if sig validation fails - but only when there is policy checking sigs.

TL;DR; is some debt here already - will add this to the list of items to revisit when improving the enrichers, but for right now recommend leaving it as.

**Review by jvdm — APPROVED** [review]
